plugins/core/core/src/software/aws/toolkits/core/utils/RemoteResourceResolver.kt

```diff
@@ -13,30 +15,62 @@ import java.time.Instant
 import java.util.UUID
 import java.util.concurrent.Callable
 import java.util.concurrent.CompletionStage
+import java.util.concurrent.atomic.AtomicBoolean
 
 interface RemoteResourceResolver {
     fun resolve(resource: RemoteResource): CompletionStage<Path>
+    fun checkForUpdates(endpoint: String, eTagProvider: ETagProvider): UpdateCheckResult
+    fun getLocalResourcePath(filename: String): Path?
 }
 interface RemoteResolveParser {
     fun canBeParsed(data: InputStream): Boolean
 }
 
+interface ETagProvider {
+    var etag: String?
+    fun updateETag(newETag: String?)
+}
+
+sealed class UpdateCheckResult {
+    data object HasUpdates : UpdateCheckResult()
+    data object NoUpdates : UpdateCheckResult()
+    data object FirstPollCheck : UpdateCheckResult()
+}
+
 class DefaultRemoteResourceResolver(
     private val urlFetcher: UrlFetcher,
     private val cacheBasePath: Path,
     private val executor: (Callable<Path>) -> CompletionStage<Path>,
 ) : RemoteResourceResolver {
+    private val isFirstPoll = AtomicBoolean(true)
 
     override fun resolve(resource: RemoteResource): CompletionStage<Path> = executor(Callable { internalResolve(resource) })
 
+    override fun getLocalResourcePath(filename: String): Path? {
+        val expectedLocation = cacheBasePath.resolve(filename)
+        return expectedLocation.existsOrNull()
+    }
+
+    override fun checkForUpdates(endpoint: String, eTagProvider: ETagProvider): UpdateCheckResult {
+        val hasETagUpdate = updateETags(eTagProvider, endpoint)
+        // for when we need to notify on first poll even when there's no new ETag
```

Comment:
it would make more sense to see if ETagProvider has a value set to determine if this were the first request